src/libfsm/determinise.c

```diff
@@ -284,109 +284,171 @@ nextnotdone(struct mapping_set *mappings, struct mapping_iter_save *sv)
 {
 	struct fsm_state *s;
 	struct state_iter it;
+	int sym, sym_min, sym_max;
+	int ret = 0;
 
-	assert(set != NULL);
-	assert(trans != NULL);
+	struct state_array outedges[FSM_SIGMA_COUNT];
+	struct hashset *outsets[FSM_SIGMA_COUNT];
 
-	for (s = state_set_first(set, &it); s != NULL; s = state_set_next(&it)) {
+	assert(fsm != NULL);
+	assert(dfa != NULL);
+	assert(curr != NULL);
+	assert(curr->closure != NULL);
+
+	sym_min = FSM_SIGMA_COUNT;
+	sym_max = -1;
+
+	memset(outedges, 0, sizeof outedges);
+	memset(outsets, 0, sizeof outsets);
+
+	for (s = state_set_first(curr->closure, &it); s != NULL; s = state_set_next(&it)) {
 		struct fsm_edge *e;
 		struct edge_iter jt;
 
 		for (e = edge_set_first(s->edges, &jt); e != NULL; e = edge_set_next(&jt)) {
-			struct fsm_state *st;
-			struct state_iter kt;
+			sym = e->symbol;
 
-			for (st = state_set_first(e->sl, &kt); st != NULL; st = state_set_next(&kt)) {
-				struct trans *p, search;
+			assert(sym >= 0);
+			assert(sym < FSM_SIGMA_COUNT);
 
-				assert(st != NULL);
+			if (sym < sym_min) { sym_min = sym; }
+			if (sym > sym_max) { sym_max = sym; }
 
-				/* Skip transitions we've already got */
-				search.c = e->symbol;
-				if ((p = trans_set_contains(trans, &search))) {
-					continue;
+			/* first edge, we just copy the states */
+			if (outedges[sym].states == NULL) {
+				if (!state_array_copy_set(&outedges[sym], e->sl)) {
+					goto finish;
 				}
-
-				p = f_malloc(fsm, sizeof *p);
-				if (p == NULL) {
-					clear_trans(fsm ,trans);
-					return 0;
+			} else {
+				struct fsm_state *st;
+				struct state_iter kt;
+
+				/* wait for a second edge to make a hash set */
+				if (outsets[sym] == NULL) {
+					size_t i;
+
+					outsets[sym] = hashset_create(hash_single_state, cmp_single_state);
+					if (outsets[sym] == NULL) {
+						goto finish;
+					}
+
+					assert(outedges[sym].states != NULL);
+					assert(outedges[sym].len > 0);
+
+					/* add states from first edge */
+					for (i=0; i < outedges[sym].len; i++) {
+						if (!hashset_add(outsets[sym], outedges[sym].states[i])) {
+							goto finish;
+						}
+					}
 				}
 
-				p->c = e->symbol;
-				p->state = st;
-
-				if (!trans_set_add(trans, p)) {
-					f_free(fsm, p);
-					clear_trans(fsm, trans);
-					return 0;
+				/* iterate over states, add to state list if they're not already in the set */
+				for (st = state_set_first(e->sl, &kt); st != NULL; st = state_set_next(&kt)) {
+					if (!hashset_contains(outsets[sym], st)) {
+						if (!state_array_add(&outedges[sym], st)) {
+							goto finish;
+						}
+					}
 				}
 			}
 		}
 	}
 
-	return 1;
-}
+	if (sym_min < 0 || sym_max >= FSM_SIGMA_COUNT) {
+		goto finish;
+	}
 
-/*
- * Return a list of all states reachable from set via the given transition.
- */
-static int
-allstatesreachableby(struct state_set *set, char c, struct state_set *sl)
-{
-	struct fsm_state *s;
-	struct state_iter it;
+	for (sym = sym_min; sym <= sym_max; sym++) {
```

Comment:
I'd `for (sym = 0; sym < UCHAR_MAX; sym++)` here rather than `<=`, for ease of grepping against the existing code elsewhere.

Reply:
That would add a lot of unnecessary iteration.  I kept track of the min/max sym out based on the observation that many states have a small number of edges that are close to each other.

Reply:
For the record, we'd misunderstood each other here - i was talking about FSM_SIGMA_COUNT vs. a character value, not about the min/max endpoints.